### PR TITLE
Postpone initialization until KotlinSourceSet.dependsOn is populated

### LINKED ISSUE #510: Build failed with Kotlin 1.9.0-Beta
https://github.com/xpathexception/MokoResourcesIssue/tree/0c74975b7cf0a6581fb1113711dd73e4a0596ec6

here problem reproduced.

android build - `./gradlew :frontend:app:buildDebug`
```
> Task :frontend:modules:features:sample:compileDebugKotlinAndroid FAILED
e: file:///Users/amikhailov/Documents/projects/sandbox/MokoResourcesIssue/frontend/modules/features/sample/build/generated/moko/commonMain/src/com/example/myapp/features/sample/sampleR.kt:12:22 Expected object 'sampleR' has no actual declaration in module <sample_debug> for JVM

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':frontend:modules:features:sample:compileDebugKotlinAndroid'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details
```
as i see `frontend/modules/features/sample/build/generated/moko/androidMain/src` not marked as source root.

web build - `./gradlew :frontend:app-web:jsBrowserDevelopmentRun`
```
> Task :frontend:modules:features:sample:compileKotlinJs FAILED
e: file:///Users/amikhailov/Documents/projects/sandbox/MokoResourcesIssue/frontend/modules/features/sample/build/generated/moko/commonMain/src/com/example/myapp/features/sample/sampleR.kt:12:22 Expected object 'sampleR' has no actual declaration in module <MokoResourcesIssue.frontend.modules.features:sample> for JS

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':frontend:modules:features:sample:compileKotlinJs'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details
```
same problem - `frontend/modules/features/sample/build/generated/moko/jsMain/src` not marked as source root

### PR DESCRIPTION
**Pull request by illarionov**
Starting with Kotlin 1.90, the `KotlinSourceSet.dependsOn` results are not guaranteed to be up to date even if this method it called from the `afterEvaluate` block.
To be sure if its work, the internal state of the Kotlin Multiplatform Plugin should be at least at the `KotlinPluginLifecycle.Stage.AfterFinaliseRefinesEdges` stage (See [defaultKotlinHierarchySetup.kt](https://github.com/JetBrains/kotlin/blob/3ff21db77f983249a07c29be4988a39370b5035a/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/hierarchy/defaultKotlinHierarchySetup.kt#L24C1)).

We need to wait for this state before using this method. Unfortunately, Kotlin Multiplatform Plugin does not provide any public API for tracking the current internal state.

In this pull request, state tracking is implemented using reflection with fallback to the calling `afterEvaluate` multiple times.

Fixes #510, #531.

As one of the possible solutions. Leave a comment if you have better ideas on how to fix this.

### CODE REVIEW

**Comment by Alex009** [comment]
hi! thanks for research! current fix looks like hack, but if it works and we will not find better solution - i will approve this request.

your research shows that now we run `configureGenerators` at `afterEvaluate` and inside read targets and sourceSets at time of `afterEvaluate` call, but at this time on Kotlin 1.9.0 sourceSets/targets not created yet? did i understand correctly?
if yes - did kotlin plugin use gradle live collections? maybe we should use `configureEach` to correctly capture moment when kotlin source set was configured and available to processing by our plugin? 

also maybe some JetBrains team find time and give feedback for us about this fix? @terrakok for example

**Comment by sellmair** [comment]
I see this MR is using reflection to get into absolutely internal APIs that can change at any given time. 
I am not in the position to tell you folks how to go forward with this issue, but I would really like to advise you not to do this. 
My prediction is that this gets you into a deep rabbit hole where you end up just being compatible with a single version of the Kotlin Gradle Plugin. 

Here is my offer: 
Lets jump into a call and figure out a 'short term' solution and a 'long term' one. 
Short term: We can find out the API that you can use for now (that we know and keep stable then)
Long term: We give you the API that you need and make it public.

File: resources-generator/src/main/kotlin/dev/icerock/gradle/utils/LifecycleAwareCodeLauncher.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ */
+internal fun Project.launchWhenEdgeRefinementFinalized(block: () -> Unit) {
+    val version = project.plugins.getPlugin(KotlinMultiplatformPluginWrapper::class).pluginVersion
+        .split(".")
+        .take(2)
+        .map(String::toInt)
+
+    @Suppress("MagicNumber")
+    when {
+        version[0] >= 1 && version[1] >= 9 -> {
```
Comment (bcmedeiros) [inline resources-generator/src/main/kotlin/dev/icerock/gradle/utils/LifecycleAwareCodeLauncher.kt:25]: `version[0] >= 1 && version[1] >= 9` returns false for 2.4, for example, we need something like `version[0] > 1 || version[0] = 1 && version[1] >= 9`

Comment (illarionov) [reply resources-generator/src/main/kotlin/dev/icerock/gradle/utils/LifecycleAwareCodeLauncher.kt]: good point, thank you!

**Comment by sellmair** [comment]
I still see some activity here, but no reply on my message. 
As I am technically leading the Kotlin Gradle Plugin for JetBrains, I really hope my concerns have been acknowledged and I would really like to repeat my offer: We can look for solutions together and collaborate.

**Comment by Alex009** [comment]
@sellmair i 100% agree with you about using reflection and internal api. i not want to use it. 
but at now i can't say we have enough api for implementation our logic or not (i not check by self it yet), so i can't now say that something should be changed in KGP. 
if kotlin.targets is live collection and target sourceSets live collection too - i think we can observe changes and detect when target/sourceSet added.
when i dive into details i return here

**Comment by illarionov** [comment]
Oh, I'm sorry, never mind, I see now that we can solve this problem with the current API without using reflection.
My main concern is that we cannot use the `KotlinSourceSet.dependsOn` getter at the Gradle configuration stage and will have to defer it to a later task exeution stages.
This method would be useful at the configuration stage in this plugin, since we register new gradle tasks based on it results.  
But I think that we can solve this by registering all tasks in advance and then disabling unnecessary ones.

**Pull request closed without merging**